go/chat/searcher.go

```diff
@@ -25,58 +25,121 @@ func NewSearcher(g *globals.Context) *Searcher {
 	}
 }
 
-func (s *Searcher) SearchRegexp(ctx context.Context, uiCh chan chat1.ChatSearchHit, conversationID chat1.ConversationID, re *regexp.Regexp, maxHits int, maxMessages int) (hits []chat1.ChatSearchHit, rlimits []chat1.RateLimit, err error) {
+func (s *Searcher) SearchRegexp(ctx context.Context, uiCh chan chat1.ChatSearchHit, conversationID chat1.ConversationID, re *regexp.Regexp,
+	maxHits, maxMessages, beforeContext, afterContext int) (hits []chat1.ChatSearchHit, rlimits []chat1.RateLimit, err error) {
 	uid := gregor1.UID(s.G().Env.GetUID().ToBytes())
 	pagination := &chat1.Pagination{Num: pageSize}
-	typs := []chat1.MessageType{chat1.MessageType_TEXT}
-	getThreadQuery := &chat1.GetThreadQuery{MessageTypes: typs}
 
-	var prev *chat1.MessageUnboxed
-	var next *chat1.MessageUnboxed
-	var searchHit chat1.ChatSearchHit
-	var rlimitsp []*chat1.RateLimit
-	numHits := 0
-	numMessages := 0
-	getPresentMsg := func(msg *chat1.MessageUnboxed) *chat1.UIMessage {
-		if msg != nil {
-			uiMsg := utils.PresentMessageUnboxed(ctx, s.G(), *msg, uid, conversationID)
-			return &uiMsg
-		}
-		return nil
+	if beforeContext >= pageSize {
+		beforeContext = pageSize - 1
 	}
-	for !pagination.Last && numHits < maxHits && numMessages < maxMessages {
-		thread, rl, err := s.G().ConvSource.Pull(ctx, conversationID, uid, getThreadQuery, pagination)
+	if afterContext >= pageSize {
+		afterContext = pageSize - 1
+	}
+
+	var rlimitsp []*chat1.RateLimit
+	// If we have to gather search result context around a pagination boundary
+	// we may have to fetch a thread and need to prevent refetching after this.
+	var prevThread, curThread *chat1.ThreadView
+	fetchThread := true
+
+	getThread := func() error {
+		prevThread = curThread
+		thread, rl, err := s.G().ConvSource.Pull(ctx, conversationID, uid, &chat1.GetThreadQuery{
+			MessageTypes: []chat1.MessageType{chat1.MessageType_TEXT},
+		}, pagination)
 		if err != nil {
-			return hits, rlimits, err
+			return err
 		}
-		pagination = thread.Pagination
+		curThread = &thread
+		pagination = curThread.Pagination
 		pagination.Num = pageSize
 		pagination.Previous = nil
 		rlimitsp = append(rlimitsp, rl...)
+		return nil
+	}
+
+	getBeforeMsgs := func(i int) ([]chat1.MessageUnboxed, error) {
```

Comment:
I find these functions extremely difficult to read, can we make the parameter list more explicit?